### ISSUE TITLE
Consolidated Zarr support could improve S3 data loading

### ISSUE THREAD

**Issue by mannreis** [Unidata/netcdf-c#2987]
Hello 👋

We've noticed the difference between reading a remote Zarr dataset [`https://...#mode=s3,zarr`] and local one [`file://....#mode=file,zarr`]:
```
$ time ncdump/ncdump -v tas file://${HOME}/${DATASET}/#mode=zarr | tail -n+2 | md5sum
abd28bc55fb9d0c25a3767a43d27110a -
real 0m0.111s
user 0m0.104s
sys 0m0.017s
```
```
$ time ncdump/ncdump -v tas https://${ENDPOINT}/${BUCKET}/${DATASET}/#mode=zarr,s3 | tail -n+2 | md5sum
abd28bc55fb9d0c25a3767a43d27110a -
real 0m9.854s
user 0m4.739s
sys 0m0.162s
```
Network overhead is expected, specially if the service imposes rate limits. But such a difference motivated me to look at the implementation behaviour. 

It seems that the approach used by netcdf is similar to the one used with Python Zarr, fetching all the metadata in advance. And for this reason the following requests are sent (in netcdf) for my the example above:
 - 4 GET Requests to list the dataset metadatafiles
 - 674 HEAD Requests to mainly fetch the size of the object to be transfered
 - 224 GET Requests to actually read the content of both metadata (223) and data/chunks (1) objects.
 
 There are 3x more HEAD than GET which can be a tinny improvement, but overall this is not much different from what Python does:
 ```Python
import zarr,s3fs,os
s3 = s3fs.S3FileSystem(endpoint_url=f'https://{os.environ["ENDPOINT"]}/',anon=True)
store = s3fs.S3Map(root=f'{os.environ["BUCKET"]}/{os.environ["DATASET"]}',s3=s3)
d = zarr.open(store,'r')
print(d.info)
 ```
 Which produces:
 - 127 GET Requests to list metadatafiles or variable names after the dataset prefix
 - 349 HEAD Requests to check for metadatafiles
 
 ----
Implementing a [consolidated access](https://zarr.readthedocs.io/en/stable/tutorial.html#consolidating-metadata) mode could improve the situation. In Python, the example above can be simplified to a single request:
- 1 GET Request to fetch the content of `/.zmetadata` (Note that not even a HEAD request is done in advance)
```python
import zarr,s3fs,os
s3 = s3fs.S3FileSystem(endpoint_url=f'https://{os.environ["ENDPOINT"]}/',anon=True)
store = s3fs.S3Map(root=f'{os.environ["BUCKET"]}/{os.environ["DATASET"]}',s3=s3)
d = zarr.open_consolidated(store)
print(d.info)
```

If this is desired perhaps it could be supported by other modes, like `file` (or even `zip`!?) as well. In that case I think it would be part of the zarr api and not a specific zmap S3 implementation.

I will try to come up with a PR for this but it would be great to have some feedback and if positive, some pointers/draft on how to support it (via `#mode=consolidated` controls? Environment? Only when build --with-consolidated-zarr?)

Thanks!

**Comment by joshmoore** [comment]
:+1: for support of Zarr v2 "consolidated". A discussion on the zarr-python team yesterday touched on how to deal with the potential differences in this respect between possible differences in the definition of consolidated between the v2 and v3 formats. The decision was to add arguments to enable the v2 consolidated _format_ to the v3 _library_, but potentially disallow those arguments when producing the v3 _format_ (since the v3 _library_ will need to support both the v2 and v3 _formats_.) `</tongue_twister>`

**Comment by DennisHeimbigner** [comment]
Sorry, I apparently missed this Issue when it was first posted.
In any case, we have always planned to support consolidated
metadata for both V2 and V3. The problem was that there appeared
to be no specification for the JSON for consolidated metadata.
Has that changed? Can you point me to that spec?
Josh's note about V3 supporting both V2 and V3 is unclear.
I get that the actual (python) library will need to read files in both V2
and V3 formats. But I do not understand this remark:
````
The decision was to add arguments to enable the v2 consolidated format to the v3 library, but potentially disallow those arguments when producing the v3 format
````
What kind of arguments are being considered?

**Comment by joshmoore** [comment]
> The problem was that there appeared
> to be no specification for the JSON for consolidated metadata.
> Has that changed? Can you point me to that spec?

No, that has not changed, but agreed that that is a difficult for the v2 format.

> What kind of arguments are being considered?

Correction. For zarr-python library v2, I should have said "methods" or "API" for activating consolidated metadata. (Those don't yet exist for zarr-python library v3.) The method *arguments* I was thinking of are in xarray: https://docs.xarray.dev/en/stable/user-guide/io.html#consolidated-metadata

**Comment by DennisHeimbigner** [comment]
Ok, I see.
So the big holdup at the moment is a JSON spec for
consolidated metadata for V2 and another for V3.

**Comment by mannreis** [comment]
Thanks for the discussion! In the meantime I've tried to just add a "caching layer" the metadata functions that would GET the .z* files to see what the difference would be [1] . I've opened #2992 but its a draft perhaps not useful on the long term.

[1]
```
$ time ncdump/ncdump -v tas https://${ENDPOINT}/${BUCKET}/${DATASET}/#mode=zarr,s3,consolidated | tail -n+2 | md5sum
abd28bc55fb9d0c25a3767a43d27110a -
real	0m0.262s
user	0m0.155s
sys	0m0.022s
```

**Comment by poidl** [comment]
wow, this looks great! Does this mean that we can use use netcdf-c
to download zarr files from an S3 server that

* does not support object listing, but
* does have a .zmetadata file?

I just pulled the merged commits and tried this:
```
ncdump -h 'https://.../#mode=zarr,s3,consolidated'
```
with a provider that does not seem to allow s3 listing, and it
fails because the first http request is still a
```
GET /bucketname?delimiter=%2F&list-type=2...
...
content-type: application/xml
```
request. I expected a GET .zmetadata instead, without prior
listing. Is this already possible with the current implementation?
I'm new to zarr and may misunderstand how to use it.

**Comment by DennisHeimbigner** [comment]
Yes, your request should work. We wiil check to see where that List2 call is originating.
Can I access that source and if so what is it?

**Comment by DennisHeimbigner** [comment]
I recall now.
We do not use list2 to get all nodes, but we do use list2 with an appropriate
set of prefixes to get the set of nodes immediately below another (typically a group node).
List2 is the only way to do this.
If you want to post the full GET line, I can run you thru it.

**Comment by mannreis** [comment]
Thanks for your interest! It's nice to see we aren't the only ones wishing to use a server that doesn't fully "speak" S3.

This issue aimed at reporting the performance impact and **not** the issue of using few S3 `list()`.

> Does this mean that we can use use netcdf-c to download zarr files from an S3 server that
> * does not support object listing, but

So I'll close it once the new release is out.

Actually, no! As you saw that are 2 places where a `list()` is still done in the current implementation, and by specifiying (note `s3`):
> ```
> ncdump -h 'https://.../#mode=zarr,s3,consolidated'
> ```
You're telling the ncdump the remote is indeed an S3-capable one.

Yet, it seems too easy to make it work without any `list()`: https://github.com/Unidata/netcdf-c/pull/3241
In my opinion it would be reasonable to drop `list()` for the case where the user specifies `NCZARR_CONSOLIDATED` / `mode=consolidated`. What is your opinion @DennisHeimbigner?

We do want to broaden the usage with a general `mode=http` that requires the consolidated metadata. But that is a different discussion you can contribute to: https://github.com/Unidata/netcdf-c/discussions/3228

**Comment by mannreis** [comment]
> If you want to post the full GET line, I can run you thru it.

Feel free to use until it stops working:
```
ncdump -v time https://swift.dkrz.de/v1/dkrz_948e7d4bbfbb445fbff5315fc433e36a/hera5/v2/hera5_P1D_hpz7.zarr#mode=zarr,s3,consolidated`
```

**Comment by poidl** [comment]
Great! I merged mannreis/zarr-csl-fully and it works indeed with my provider!

Thanks also for posting the link for testing! I was hesitant to post the url because the provider distributes an API client that requires a login with credentials. Even though many of the assets are unprotected, there is apparently some user tracking attempted via query strings.

Regarding the test url, I noticed that the server response is `412 Precondition Failed` whereas mine was `403 Forbidden`. But yes, the end result is the same.

I don't know much about s3 or netcdf, but regarding the discussion about different modes and possible automated discovery of which mode to apply, it occurred to me that there may come a time when my provider decides to properly lock down access and allow only s3 signed/autheticated requests to GET objects. This would be a case where 'mode=s3,consolidated' would be required (including your patch in case listing is forbidden) as opposed to e.g. 'mode=zoh' (zarr over http), because the latter does no s3 request signing. Hmm, maybe there is a clause in the s3 spec saying that it MUST be discoverable by a client whether listing is allowed before making the actual listing request? Not sure if that would help.

Anyways, for my personal use case everything is resolved! THANKS AGAIN.

**Comment by WardF** [comment]
Closed by #3241 Thank you @mannreis!
